File: SWEA/D3/2806. N－Queen/N－Queen.cpp
#include <iostream>
#include <vector>
using namespace std;

int N, count;

bool safe(int col, int row, vector<int>& v){
    for(int prev_row = 0; prev_row < row; prev_row++){
        int prev_col = v[prev_row];
        if(prev_col == col || abs(prev_row - row) == abs(prev_col - col)){
            return false;
        }
    }
    return true;
}

void dfs(int row, vector<int>& v){
    if(row == N){
        count++;
        return;
    }
    for(int col = 0; col < N; col++){
        if(safe(col, row, v)){
            v[row] = col;
            dfs(row + 1, v);
        }
    }
}

int main(){
    int T;
    cin >> T;
    for(int t = 1; t <= T; t++){
        cin >> N;
        count = 0;
        vector<int> v(N, -1);

        dfs(0, v);

        cout << "#" << t << ' ' << count << '\n';
    }
    return 0;
}

File: SWEA/D3/2806. N－Queen/N－Queen.java
import java.util.Scanner;

public class Solution {
    static int ans, N;
    static int[][] board;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        for(int test_case = 1; test_case <= T; test_case++){
            N = sc.nextInt();
            board = new int[N][N];
            ans = 0;
            dfs(0);
            System.out.println("#" + test_case + " " + ans);
        }
    }

    static void dfs(int row){
        if(row == N){
            ans++;
            return;
        }
        
        for(int col = 0; col < N; col++){
            if(can(row, col)){
                board[row][col] = 1;
                dfs(row + 1);
                board[row][col] = 0;
            }
        }
    }

    static boolean can(int row, int col){
        for(int i = 0; i < row; i++){         //row 검사
            if(board[i][col] == 1){
                return false;
            }
        }

        for(int i = 1; row - i >= 0 && col - i >= 0; i++){      //좌측대각 검사
            if(board[row - i][col - i] == 1){
                return false;
            }
        }

        for(int i = 1; row - i >= 0 && col + i < N; i++){       //우측대각 검사
            if(board[row - i][col + i] == 1){
                return false;
            }
        }

        return true;
    }
}
